src/clj/game/core/moving.clj

```diff
@@ -518,6 +514,24 @@
         (when (in-hand? moved-b) (add-to-currently-drawing state b-side moved-b)))
       [(get-card state moved-a) (get-card state moved-b)])))
 
+(defn swap-cards-async
+  "Swaps two cards when one or both aren't installed"
+  ([state side a b] (swap-cards-async state side nil a b))
+  ([state side eid a b]
+   (let [async-result (swap-cards state side a b)
+         moved-a (first async-result)
+         moved-b (second async-result)
+         install-event (= 1 (count (filter installed? [moved-a moved-b])))]
+     ;; todo - we might need behaviour for runner swap installs down the line, depending on future cards
+     ;; that's a problem for another day
+     (if (and install-event (= :corp side))
+       (trigger-event-sync
+         state side eid :corp-install
+         {:card (get-card state (if (installed? moved-a) moved-a moved-b))
+          :install-state (:install-state (card-def (if (installed? moved-a) moved-a moved-b)))})
+       (effect-completed eid))
+     async-result)))
```

Comment:
Sadly, can't just return the async-result, you need to use `complete-with-result`: `(if (and ...) (trigger ...) (complete-with-result state side eid async-result)`

however, you also can't use `trigger-event-sync`, lol. It's been converted to a checkpoint, so you need to say (this is a mouthful, my apologies):

```clojure
(if (and install-event (= :corp side))
  (do (queue-event state :corp-install {:card ... :install-state ...})
      (wait-for (checkpoint state nil (make-eid state eid))
                (complete-with-result state side eid async-result)))
  (complete-with-result state side eid async-result))
```